### PR TITLE
[Reasoning] Add thinking budget support 

### PR DESCRIPTION
**Pull request by rishitdholakia13**
…e methods

## Essential Elements of an Effective PR Description Checklist
- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

## Purpose
This PR adds thinking budget support for reasoning models. This support has been request as per #15418  and helps build upon changes from #20859 . This PR helps in the following:

- Adds support for models that support speculative and non-speculative decoding
- Avoids the need to use logits processor. 
- Uses thinking tokens from reasoningparser
- Allows free thinking, if thinking_bedget is set to -1, then we allow the model think freely without any thinking budget

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Does it allow setting custom thinking-budget-exceeded message? E.g. so that if thinking budget is exceeded, vllm would append to the current response a custom message: `</think>Thinking was interrupted. Now let's formulate the final response:`? (which would also allow the client to detect that thinking-budget was exceeded)
